File: internal/engine/engine.go
package engine

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/mortenaho/workflowengine/internal/domain"
	"github.com/mortenaho/workflowengine/internal/identity"
	"github.com/mortenaho/workflowengine/internal/store"
)

type Option func(*Engine)

func WithClock(clock func() time.Time) Option {
	return func(e *Engine) { e.clock = clock }
}

type Engine struct {
	store store.Store
	dir   identity.Directory
	clock func() time.Time
}

func New(s store.Store, d identity.Directory, opts ...Option) *Engine {
	e := &Engine{store: s, dir: d, clock: time.Now}
	for _, opt := range opts {
		opt(e)
	}
	return e
}

func (e *Engine) now() time.Time { return e.clock().UTC() }

func (e *Engine) Register(ctx context.Context, key, name string) (*domain.Definition, error) {
	key = strings.TrimSpace(key)
	if key == "" {
		return nil, fmt.Errorf("%w: process key is required", domain.ErrInvalid)
	}
	if existing, err := e.store.GetDefinitionByKey(ctx, domain.TenantID(ctx), key); err == nil {
		if name != "" && existing.Name != name {
			existing.Name = name
			_ = e.store.SaveDefinition(ctx, existing)
		}
		return existing, nil
	}
	if name == "" {
		name = key
	}
	def := &domain.Definition{
		ID:        domain.NewID(),
		TenantID:  domain.TenantID(ctx),
		Key:       key,
		Name:      name,
		CreatedAt: e.now(),
	}
	if err := e.store.SaveDefinition(ctx, def); err != nil {
		return nil, err
	}
	return def, nil
}

func (e *Engine) GetDefinitionByKey(ctx context.Context, key string) (*domain.Definition, error) {
	return e.store.GetDefinitionByKey(ctx, domain.TenantID(ctx), key)
}

func (e *Engine) Start(ctx context.Context, processKey, initiator string, params domain.Vars) (*domain.StartResult, error) {
	processKey = strings.TrimSpace(processKey)
	initiator = strings.TrimSpace(initiator)
	if processKey == "" {
		return nil, fmt.Errorf("%w: processKey is required", domain.ErrInvalid)
	}
	if initiator == "" {
		return nil, fmt.Errorf("%w: initiator is required", domain.ErrInvalid)
	}
	def, err := e.Register(ctx, processKey, "")
	if err != nil {
		return nil, err
	}
	now := e.now()
	inst := &domain.ProcessInstance{
		ID:            domain.NewID(),
		TenantID:      domain.TenantID(ctx),
		DefinitionID:  def.ID,
		DefinitionKey: def.Key,
		Status:        domain.InstanceRunning,
		Parameters:    params.Clone(),
		StartedBy:     initiator,
		CreatedAt:     now,
		UpdatedAt:     now,
	}
	if err := e.store.CreateInstance(ctx, inst); err != nil {
		return nil, err
	}
	return &domain.StartResult{DefinitionKey: def.Key, InstanceID: inst.ID}, nil
}

func (e *Engine) GetInstance(ctx context.Context, id string) (*domain.ProcessInstance, error) {
	inst, err := e.store.GetInstance(ctx, id)
	if err != nil {
		return nil, err
	}
	if domain.NormalizeTenant(inst.TenantID) != domain.TenantID(ctx) {
		return nil, domain.ErrForbiddenTenant
	}
	return inst, nil
}

func (e *Engine) ListByProcessKey(ctx context.Context, processKey string) (*domain.ProcessList, error) {
	processKey = strings.TrimSpace(processKey)
	if processKey == "" {
		return nil, fmt.Errorf("%w: processKey is required", domain.ErrInvalid)
	}
	roots, err := e.store.ListRootInstances(ctx, domain.TenantID(ctx), processKey)
	if err != nil {
		return nil, err
	}
	list := &domain.ProcessList{
		ProcessKey: processKey,
		Instances:  make([]*domain.ProcessInstanceDetail, 0, len(roots)),
	}
	for _, inst := range roots {
		if domain.NormalizeTenant(inst.TenantID) != domain.TenantID(ctx) {
			continue
		}
		tasks, err := e.store.ListTasks(ctx, domain.TaskFilter{
			InstanceID: inst.ID,
			TenantID:   domain.TenantID(ctx),
		})
		if err != nil {
			return nil, err
		}
		if tasks == nil {
			tasks = []*domain.Task{}
		}
		detail := detailFrom(inst, processKey, tasks)
		list.Instances = append(list.Instances, detail)
	}
	list.Total = len(list.Instances)
	return list, nil
}

func detailFrom(inst *domain.ProcessInstance, processKey string, tasks []*domain.Task) *domain.ProcessInstanceDetail {
	d := &domain.ProcessInstanceDetail{
		InstanceID:    inst.ID,
		ProcessKey:    processKey,
		DefinitionKey: inst.DefinitionKey,
		Initiator:     inst.StartedBy,
		Status:        inst.Status,
		Parameters:    inst.Parameters.Clone(),
		CreatedAt:     inst.CreatedAt,
		UpdatedAt:     inst.UpdatedAt,
		Tasks:         tasks,
		TaskTotal:     len(tasks),
	}
	for _, t := range tasks {
		if t.Status == domain.TaskDone {
			d.TasksCompleted++
		} else {
			d.TasksOpen++
		}
	}
	d.AllTasksCompleted = d.TaskTotal > 0 && d.TasksOpen == 0
	return d
}

func (e *Engine) Refer(ctx context.Context, actor string, in domain.ReferInput) (*domain.ReferResult, error) {
	actor = strings.TrimSpace(actor)
	if actor == "" {
		return nil, fmt.Errorf("%w: actor is required", domain.ErrInvalid)
	}
	defKey := strings.TrimSpace(in.DefinitionKey)
	parentID := strings.TrimSpace(in.ParentInstanceID)
	var parent *domain.ProcessInstance
	if parentID != "" {
		var err error
		parent, err = e.GetInstance(ctx, parentID)
		if err != nil {
			return nil, err
		}
		if defKey == "" {
			defKey = parent.DefinitionKey
		} else if defKey != parent.DefinitionKey {
			return nil, fmt.Errorf("%w: definitionKey does not match parent process", domain.ErrInvalid)
		}
	}
	if defKey == "" {
		return nil, fmt.Errorf("%w: definitionKey is required", domain.ErrInvalid)
	}
	def, err := e.store.GetDefinitionByKey(ctx, domain.TenantID(ctx), defKey)
	if err != nil {
		return nil, fmt.Errorf("%w: unknown definition %s (start the process first)", err, defKey)
	}

	kind := in.ToKind
	ids := uniqueIDs(in.ToIDs)
	if in.ToID != "" {
		ids = uniqueIDs(append(ids, in.ToID))
	}
	switch kind {
	case domain.AssigneeUser:
		if len(ids) != 1 {
			return nil, fmt.Errorf("%w: user referral needs exactly one id", domain.ErrInvalid)
		}
	case domain.AssigneeGroup:
		if len(ids) != 1 {
			return nil, fmt.Errorf("%w: group referral needs exactly one id", domain.ErrInvalid)
		}
		members, err := e.dir.GroupMembers(ctx, ids[0])
		if err != nil {
			return nil, err
		}
		if len(members) == 0 {
			return nil, domain.ErrEmptyGroup
		}
	case domain.AssigneeUsers:
		if len(ids) == 0 {
			return nil, fmt.Errorf("%w: users referral needs at least one id", domain.ErrInvalid)
		}
		kind = domain.AssigneeUser
	default:
		return nil, fmt.Errorf("%w: to.kind must be user, group, or users", domain.ErrInvalid)
	}

	now := e.now()
	inst := &domain.ProcessInstance{
		ID:               domain.NewID(),
		TenantID:         domain.TenantID(ctx),
		DefinitionID:     def.ID,
		DefinitionKey:    def.Key,
		ParentInstanceID: parentID,
		Status:           domain.InstanceRunning,
		Parameters:       in.Parameters.Clone(),
		StartedBy:        actor,
		CreatedAt:        now,
		UpdatedAt:        now,
	}
	if err := e.store.CreateInstance(ctx, inst); err != nil {
		return nil, err
	}

	tasks := make([]*domain.Task, 0, len(ids))
	for _, id := range ids {
		t := &domain.Task{
			ID:               domain.NewID(),
			TenantID:         domain.TenantID(ctx),
			InstanceID:       inst.ID,
			ParentInstanceID: parentID,
			DefinitionKey:    def.Key,
			Title:            in.Title,
			AssigneeKind:     kind,
			AssigneeID:       id,
			AssignedBy:       actor,
			Status:           domain.TaskOpen,
			CreatedAt:        now,
			UpdatedAt:        now,
		}
		if err := e.store.SaveTask(ctx, t); err != nil {
			return nil, err
		}
		tasks = append(tasks, t)
	}
	out := &domain.ReferResult{
		InstanceID:    inst.ID,
		DefinitionKey: def.Key,
		Tasks:         tasks,
	}
	if len(tasks) == 1 {
		out.Task = tasks[0]
	}
	return out, nil
}

func uniqueIDs(ids []string) []string {
	seen := make(map[string]struct{}, len(ids))
	out := make([]string, 0, len(ids))
	for _, id := range ids {
		id = strings.TrimSpace(id)
		if id == "" {
			continue
		}
		if _, ok := seen[id]; ok {
			continue
		}
		seen[id] = struct{}{}
		out = append(out, id)
	}
	return out
}

func (e *Engine) PendingTasks(ctx context.Context, userID, groupID string) ([]*domain.Task, error) {
	userID = strings.TrimSpace(userID)
	groupID = strings.TrimSpace(groupID)
	if userID == "" && groupID == "" {
		return nil, fmt.Errorf("%w: user or group is required", domain.ErrInvalid)
	}
	filter := domain.TaskFilter{
		Status:   domain.TaskOpen,
		TenantID: domain.TenantID(ctx),
	}
	if groupID != "" {
		filter.Status = ""
		filter.GroupID = groupID
		filter.Statuses = []domain.TaskStatus{domain.TaskOpen, domain.TaskClaimed}
		tasks, err := e.store.ListTasks(ctx, filter)
		if err != nil {
			return nil, err
		}
		if tasks == nil {
			tasks = []*domain.Task{}
		}
		return tasks, nil
	}
	filter.UserID = userID
	filter.Status = domain.TaskOpen
	groups, err := e.dir.UserGroups(ctx, userID)
	if err != nil {
		return nil, err
	}
	filter.GroupIDs = groups
	open, err := e.store.ListTasks(ctx, filter)
	if err != nil {
		return nil, err
	}
	claimed, err := e.store.ListTasks(ctx, domain.TaskFilter{
		ClaimedBy: userID,
		Status:    domain.TaskClaimed,
		TenantID:  domain.TenantID(ctx),
	})
	if err != nil {
		return nil, err
	}
	return mergeTasks(open, claimed), nil
}

func mergeTasks(parts ...[]*domain.Task) []*domain.Task {
	seen := make(map[string]struct{})
	out := make([]*domain.Task, 0)
	for _, list := range parts {
		for _, t := range list {
			if t == nil {
				continue
			}
			if _, ok := seen[t.ID]; ok {
				continue
			}
			seen[t.ID] = struct{}{}
			out = append(out, t)
		}
	}
	return out
}

func (e *Engine) GetTask(ctx context.Context, id string) (*domain.Task, error) {
	t, err := e.store.GetTask(ctx, id)
	if err != nil {
		return nil, err
	}
	if domain.NormalizeTenant(t.TenantID) != domain.TenantID(ctx) {
		return nil, domain.ErrForbiddenTenant
	}
	return t, nil
}

func (e *Engine) ListTasksByInstance(ctx context.Context, instanceID string) ([]*domain.Task, error) {
	if _, err := e.GetInstance(ctx, instanceID); err != nil {
		return nil, err
	}
	tasks, err := e.store.ListTasks(ctx, domain.TaskFilter{
		InstanceID: instanceID,
		TenantID:   domain.TenantID(ctx),
	})
	if err != nil {
		return nil, err
	}
	if tasks == nil {
		tasks = []*domain.Task{}
	}
	return tasks, nil
}

func (e *Engine) Completion(ctx context.Context, instanceID string) (*domain.Completion, error) {
	if _, err := e.GetInstance(ctx, instanceID); err != nil {
		return nil, err
	}
	tasks, err := e.store.ListTasks(ctx, domain.TaskFilter{
		InstanceID: instanceID,
		TenantID:   domain.TenantID(ctx),
	})
	if err != nil {
		return nil, err
	}
	return completionOf(instanceID, tasks), nil
}

func completionOf(instanceID string, tasks []*domain.Task) *domain.Completion {
	if tasks == nil {
		tasks = []*domain.Task{}
	}
	owned := make([]*domain.Task, 0, len(tasks))
	for _, t := range tasks {
		if t.InstanceID == instanceID {
			owned = append(owned, t)
		}
	}
	c := &domain.Completion{InstanceID: instanceID, Tasks: owned, Total: len(owned)}
	for _, t := range owned {
		if t.Status == domain.TaskDone {
			c.Completed++
		} else {
			c.Open++
		}
	}
	c.AllCompleted = c.Total > 0 && c.Open == 0 && c.Completed == c.Total
	return c
}

func (e *Engine) ClaimTask(ctx context.Context, taskID, actor string) (*domain.Task, error) {
	actor = strings.TrimSpace(actor)
	if actor == "" {
		return nil, fmt.Errorf("%w: actor is required", domain.ErrInvalid)
	}
	task, err := e.GetTask(ctx, taskID)
	if err != nil {
		return nil, err
	}
	if err := e.canAct(ctx, task, actor); err != nil {
		return nil, err
	}
	if task.Status == domain.TaskClaimed {
		return nil, domain.ErrAlreadyClaimed
	}
	now := e.now()
	return e.store.TransitionTask(ctx, taskID, []domain.TaskStatus{domain.TaskOpen}, func(t *domain.Task) error {
		t.Status = domain.TaskClaimed
		t.ClaimedBy = actor
		t.UpdatedAt = now
		return nil
	})
}

func (e *Engine) UnclaimTask(ctx context.Context, taskID, actor string) (*domain.Task, error) {
	actor = strings.TrimSpace(actor)
	if actor == "" {
		return nil, fmt.Errorf("%w: actor is required", domain.ErrInvalid)
	}
	task, err := e.GetTask(ctx, taskID)
	if err != nil {
		return nil, err
	}
	if task.Status != domain.TaskClaimed {
		return nil, domain.ErrNotClaimed
	}
	if task.ClaimedBy != actor {
		return nil, domain.ErrForbidden
	}
	now := e.now()
	return e.store.TransitionTask(ctx, taskID, []domain.TaskStatus{domain.TaskClaimed}, func(t *domain.Task) error {
		t.Status = domain.TaskOpen
		t.ClaimedBy = ""
		t.UpdatedAt = now
		return nil
	})
}

func (e *Engine) CompleteTask(ctx context.Context, taskID, actor, note string, params domain.Vars) (*domain.CompleteResult, error) {
	actor = strings.TrimSpace(actor)
	if actor == "" {
		return nil, fmt.Errorf("%w: actor is required", domain.ErrInvalid)
	}
	task, err := e.GetTask(ctx, taskID)
	if err != nil {
		return nil, err
	}
	if err := e.canComplete(ctx, task, actor); err != nil {
		return nil, err
	}
	now := e.now()
	updated, err := e.store.TransitionTask(ctx, taskID, []domain.TaskStatus{domain.TaskOpen, domain.TaskClaimed}, func(t *domain.Task) error {
		if t.AssigneeKind == domain.AssigneeGroup && t.ClaimedBy != actor {
			return domain.ErrNotClaimed
		}
		t.Status = domain.TaskDone
		t.Note = note
		if t.ClaimedBy == "" {
			t.ClaimedBy = actor
		}
		t.UpdatedAt = now
		t.CompletedAt = &now
		return nil
	})
	if err != nil {
		return nil, err
	}
	inst, err := e.GetInstance(ctx, updated.InstanceID)
	if err != nil {
		return nil, err
	}
	if len(params) > 0 {
		inst.Parameters = inst.Parameters.Merge(params)
		inst.UpdatedAt = now
		if err := e.store.UpdateInstance(ctx, inst); err != nil {
			return nil, err
		}
	}
	comp, err := e.Completion(ctx, updated.InstanceID)
	if err != nil {
		return nil, err
	}
	if comp.AllCompleted && inst.Status != domain.InstanceCompleted {
		inst.Status = domain.InstanceCompleted
		inst.UpdatedAt = now
		if err := e.store.UpdateInstance(ctx, inst); err != nil {
			return nil, err
		}
	}
	return &domain.CompleteResult{Task: updated, Completion: comp}, nil
}

func (e *Engine) canAct(ctx context.Context, task *domain.Task, actor string) error {
	switch task.AssigneeKind {
	case domain.AssigneeUser:
		if task.AssigneeID != actor {
			return domain.ErrForbidden
		}
	case domain.AssigneeGroup:
		ok, err := identity.IsMember(ctx, e.dir, actor, task.AssigneeID)
		if err != nil {
			return err
		}
		if !ok {
			return domain.ErrForbidden
		}
	default:
		return domain.ErrForbidden
	}
	return nil
}

func (e *Engine) canComplete(ctx context.Context, task *domain.Task, actor string) error {
	if err := e.canAct(ctx, task, actor); err != nil {
		return err
	}
	if task.AssigneeKind == domain.AssigneeGroup {
		if task.Status != domain.TaskClaimed || task.ClaimedBy != actor {
			return domain.ErrNotClaimed
		}
	}
	if task.Status == domain.TaskClaimed && task.ClaimedBy != actor {
		return domain.ErrForbidden
	}
	return nil
}

File: src/WorkflowEngine/Engine.cs
namespace WorkflowEngine;

public sealed class Engine
{
    private readonly IStore _store;
    private readonly IDirectory _dir;
    private readonly Func<DateTime> _clock;

    public Engine(IStore store, IDirectory directory, Func<DateTime>? clock = null)
    {
        _store = store;
        _dir = directory;
        _clock = clock ?? (() => DateTime.UtcNow);
    }

    private DateTime Now()
    {
        var t = _clock();
        return t.Kind == DateTimeKind.Utc ? t : t.ToUniversalTime();
    }

    public async Task<Definition> Register(string key, string name, CancellationToken cancellationToken = default)
    {
        key = key.Trim();
        if (key.Length == 0)
            throw EngineException.Invalid("process key is required");
        var existing = await _store.GetDefinitionByKey(TenantContext.Id, key, cancellationToken);
        if (existing is not null)
        {
            if (name.Length > 0 && existing.Name != name)
            {
                existing.Name = name;
                await _store.SaveDefinition(existing, cancellationToken);
            }
            return existing;
        }
        if (name.Length == 0)
            name = key;
        var def = new Definition
        {
            Id = Ids.New(),
            TenantId = TenantContext.Id,
            Key = key,
            Name = name,
            CreatedAt = Now(),
        };
        await _store.SaveDefinition(def, cancellationToken);
        return def;
    }

    public async Task<Definition> GetDefinitionByKey(string key, CancellationToken cancellationToken = default)
    {
        var def = await _store.GetDefinitionByKey(TenantContext.Id, key, cancellationToken);
        return def ?? throw EngineException.NotFound();
    }

    public Task<Definition> LatestDefinition(string key, CancellationToken cancellationToken = default)
        => GetDefinitionByKey(key, cancellationToken);

    public async Task<StartResult> Start(string processKey, string initiator, Dictionary<string, object?>? parameters = null, CancellationToken cancellationToken = default)
    {
        processKey = processKey.Trim();
        initiator = initiator.Trim();
        if (processKey.Length == 0)
            throw EngineException.Invalid("processKey is required");
        if (initiator.Length == 0)
            throw EngineException.Invalid("initiator is required");
        var def = await Register(processKey, "", cancellationToken);
        var now = Now();
        var inst = new ProcessInstance
        {
            Id = Ids.New(),
            TenantId = TenantContext.Id,
            DefinitionId = def.Id,
            DefinitionKey = def.Key,
            Status = InstanceStatus.Running,
            Parameters = VarsUtil.Clone(parameters),
            StartedBy = initiator,
            CreatedAt = now,
            UpdatedAt = now,
        };
        await _store.CreateInstance(inst, cancellationToken);
        return new StartResult { DefinitionKey = def.Key, InstanceId = inst.Id };
    }

    public async Task<ProcessInstance> GetInstance(string id, CancellationToken cancellationToken = default)
    {
        var inst = await _store.GetInstance(id, cancellationToken);
        if (TenantContext.Normalize(inst.TenantId) != TenantContext.Id)
            throw EngineException.ForbiddenTenant();
        return inst;
    }

    public async Task<ProcessList> ListByProcessKey(string processKey, CancellationToken cancellationToken = default)
    {
        processKey = processKey.Trim();
        if (processKey.Length == 0)
            throw EngineException.Invalid("processKey is required");
        var roots = await _store.ListRootInstances(TenantContext.Id, processKey, cancellationToken);
        var list = new ProcessList { ProcessKey = processKey, Instances = [] };
        foreach (var inst in roots)
        {
            if (TenantContext.Normalize(inst.TenantId) != TenantContext.Id)
                continue;
            var tasks = await _store.ListTasks(new TaskFilter
            {
                InstanceId = inst.Id,
                TenantId = TenantContext.Id,
            }, cancellationToken);
            list.Instances.Add(DetailFrom(inst, processKey, tasks));
        }
        list.Total = list.Instances.Count;
        return list;
    }

    private static ProcessInstanceDetail DetailFrom(ProcessInstance inst, string processKey, IReadOnlyList<WorkflowTask> tasks)
    {
        var d = new ProcessInstanceDetail
        {
            InstanceId = inst.Id,
            ProcessKey = processKey,
            DefinitionKey = inst.DefinitionKey,
            Initiator = inst.StartedBy,
            Status = inst.Status,
            Parameters = VarsUtil.Clone(inst.Parameters),
            CreatedAt = inst.CreatedAt,
            UpdatedAt = inst.UpdatedAt,
            Tasks = tasks.ToList(),
            TaskTotal = tasks.Count,
        };
        foreach (var t in tasks)
        {
            if (t.Status == TaskStatus.Done)
                d.TasksCompleted++;
            else
                d.TasksOpen++;
        }
        d.AllTasksCompleted = d.TaskTotal > 0 && d.TasksOpen == 0;
        return d;
    }

    public async Task<ReferResult> Refer(string actor, ReferInput input, CancellationToken cancellationToken = default)
    {
        actor = actor.Trim();
        if (actor.Length == 0)
            throw EngineException.Invalid("actor is required");
        var defKey = input.DefinitionKey.Trim();
        var parentId = input.ParentInstanceId.Trim();
        if (parentId.Length > 0)
        {
            var parent = await GetInstance(parentId, cancellationToken);
            if (defKey.Length == 0)
                defKey = parent.DefinitionKey;
            else if (defKey != parent.DefinitionKey)
                throw EngineException.Invalid("definitionKey does not match parent process");
        }
        if (defKey.Length == 0)
            throw EngineException.Invalid("definitionKey is required");
        var def = await _store.GetDefinitionByKey(TenantContext.Id, defKey, cancellationToken)
                  ?? throw EngineException.NotFound($"unknown definition {defKey} (start the process first)");

        var kind = input.ToKind;
        var ids = UniqueIds(input.ToIds);
        if (input.ToId.Length > 0)
            ids = UniqueIds(ids.Append(input.ToId));
        switch (kind)
        {
            case AssigneeKind.User:
                if (ids.Count != 1)
                    throw EngineException.Invalid("user referral needs exactly one id");
                break;
            case AssigneeKind.Group:
                if (ids.Count != 1)
                    throw EngineException.Invalid("group referral needs exactly one id");
                var members = await _dir.GroupMembers(ids[0], cancellationToken);
                if (members.Count == 0)
                    throw EngineException.EmptyGroup();
                break;
            case AssigneeKind.Users:
                if (ids.Count == 0)
                    throw EngineException.Invalid("users referral needs at least one id");
                kind = AssigneeKind.User;
                break;
            default:
                throw EngineException.Invalid("to.kind must be user, group, or users");
        }

        var now = Now();
        var inst = new ProcessInstance
        {
            Id = Ids.New(),
            TenantId = TenantContext.Id,
            DefinitionId = def.Id,
            DefinitionKey = def.Key,
            ParentInstanceId = parentId,
            Status = InstanceStatus.Running,
            Parameters = VarsUtil.Clone(input.Parameters),
            StartedBy = actor,
            CreatedAt = now,
            UpdatedAt = now,
        };
        await _store.CreateInstance(inst, cancellationToken);

        var tasks = new List<WorkflowTask>(ids.Count);
        foreach (var id in ids)
        {
            var t = new WorkflowTask
            {
                Id = Ids.New(),
                TenantId = TenantContext.Id,
                InstanceId = inst.Id,
                ParentInstanceId = parentId,
                DefinitionKey = def.Key,
                Title = input.Title,
                AssigneeKind = kind,
                AssigneeId = id,
                AssignedBy = actor,
                Status = TaskStatus.Open,
                CreatedAt = now,
                UpdatedAt = now,
            };
            await _store.SaveTask(t, cancellationToken);
            tasks.Add(t);
        }
        var result = new ReferResult
        {
            InstanceId = inst.Id,
            DefinitionKey = def.Key,
            Tasks = tasks,
        };
        if (tasks.Count == 1)
            result.Task = tasks[0];
        return result;
    }

    private static List<string> UniqueIds(IEnumerable<string>? ids)
    {
        var seen = new HashSet<string>();
        var outList = new List<string>();
        if (ids is null)
            return outList;
        foreach (var raw in ids)
        {
            var id = raw.Trim();
            if (id.Length == 0 || !seen.Add(id))
                continue;
            outList.Add(id);
        }
        return outList;
    }

    public async Task<IReadOnlyList<WorkflowTask>> PendingTasks(string userId, string groupId, CancellationToken cancellationToken = default)
    {
        userId = userId.Trim();
        groupId = groupId.Trim();
        if (userId.Length == 0 && groupId.Length == 0)
            throw EngineException.Invalid("user or group is required");
        if (groupId.Length > 0)
        {
            var tasks = await _store.ListTasks(new TaskFilter
            {
                TenantId = TenantContext.Id,
                GroupId = groupId,
                Statuses = [TaskStatus.Open, TaskStatus.Claimed],
            }, cancellationToken);
            return tasks;
        }
        var groups = await _dir.UserGroups(userId, cancellationToken);
        var open = await _store.ListTasks(new TaskFilter
        {
            UserId = userId,
            Status = TaskStatus.Open,
            TenantId = TenantContext.Id,
            GroupIds = groups,
        }, cancellationToken);
        var claimed = await _store.ListTasks(new TaskFilter
        {
            ClaimedBy = userId,
            Status = TaskStatus.Claimed,
            TenantId = TenantContext.Id,
        }, cancellationToken);
        return MergeTasks(open, claimed);
    }

    private static List<WorkflowTask> MergeTasks(params IReadOnlyList<WorkflowTask>[] parts)
    {
        var seen = new HashSet<string>();
        var outList = new List<WorkflowTask>();
        foreach (var list in parts)
        {
            foreach (var t in list)
            {
                if (!seen.Add(t.Id))
                    continue;
                outList.Add(t);
            }
        }
        return outList;
    }

    public async Task<WorkflowTask> GetTask(string id, CancellationToken cancellationToken = default)
    {
        var t = await _store.GetTask(id, cancellationToken);
        if (TenantContext.Normalize(t.TenantId) != TenantContext.Id)
            throw EngineException.ForbiddenTenant();
        return t;
    }

    public async Task<IReadOnlyList<WorkflowTask>> ListTasksByInstance(string instanceId, CancellationToken cancellationToken = default)
    {
        _ = await GetInstance(instanceId, cancellationToken);
        return await _store.ListTasks(new TaskFilter
        {
            InstanceId = instanceId,
            TenantId = TenantContext.Id,
        }, cancellationToken);
    }

    public async Task<Completion> Completion(string instanceId, CancellationToken cancellationToken = default)
    {
        _ = await GetInstance(instanceId, cancellationToken);
        var tasks = await _store.ListTasks(new TaskFilter
        {
            InstanceId = instanceId,
            TenantId = TenantContext.Id,
        }, cancellationToken);
        return CompletionOf(instanceId, tasks);
    }

    private static Completion CompletionOf(string instanceId, IReadOnlyList<WorkflowTask> tasks)
    {
        var owned = tasks.Where(t => t.InstanceId == instanceId).ToList();
        var c = new Completion { InstanceId = instanceId, Tasks = owned, Total = owned.Count };
        foreach (var t in owned)
        {
            if (t.Status == TaskStatus.Done)
                c.Completed++;
            else
                c.Open++;
        }
        c.AllCompleted = c.Total > 0 && c.Open == 0 && c.Completed == c.Total;
        return c;
    }

    public async Task<WorkflowTask> ClaimTask(string taskId, string actor, CancellationToken cancellationToken = default)
    {
        actor = actor.Trim();
        if (actor.Length == 0)
            throw EngineException.Invalid("actor is required");
        var task = await GetTask(taskId, cancellationToken);
        await CanAct(task, actor, cancellationToken);
        if (task.Status == TaskStatus.Claimed)
            throw EngineException.AlreadyClaimed();
        var now = Now();
        return await _store.TransitionTask(taskId, [TaskStatus.Open], t =>
        {
            t.Status = TaskStatus.Claimed;
            t.ClaimedBy = actor;
            t.UpdatedAt = now;
        }, cancellationToken);
    }

    public async Task<WorkflowTask> UnclaimTask(string taskId, string actor, CancellationToken cancellationToken = default)
    {
        actor = actor.Trim();
        if (actor.Length == 0)
            throw EngineException.Invalid("actor is required");
        var task = await GetTask(taskId, cancellationToken);
        if (task.Status != TaskStatus.Claimed)
            throw EngineException.NotClaimed();
        if (task.ClaimedBy != actor)
            throw EngineException.Forbidden();
        var now = Now();
        return await _store.TransitionTask(taskId, [TaskStatus.Claimed], t =>
        {
            t.Status = TaskStatus.Open;
            t.ClaimedBy = "";
            t.UpdatedAt = now;
        }, cancellationToken);
    }

    public async Task<CompleteResult> CompleteTask(string taskId, string actor, string note, Dictionary<string, object?>? parameters = null, CancellationToken cancellationToken = default)
    {
        actor = actor.Trim();
        if (actor.Length == 0)
            throw EngineException.Invalid("actor is required");
        var task = await GetTask(taskId, cancellationToken);
        await CanComplete(task, actor, cancellationToken);
        var now = Now();
        var updated = await _store.TransitionTask(taskId, [TaskStatus.Open, TaskStatus.Claimed], t =>
        {
            if (t.AssigneeKind == AssigneeKind.Group && t.ClaimedBy != actor)
                throw EngineException.NotClaimed();
            t.Status = TaskStatus.Done;
            t.Note = note;
            if (t.ClaimedBy.Length == 0)
                t.ClaimedBy = actor;
            t.UpdatedAt = now;
            t.CompletedAt = now;
        }, cancellationToken);
        var inst = await GetInstance(updated.InstanceId, cancellationToken);
        if (parameters is { Count: > 0 })
        {
            inst.Parameters = VarsUtil.Merge(inst.Parameters, parameters);
            inst.UpdatedAt = now;
            await _store.UpdateInstance(inst, cancellationToken);
        }
        var comp = await Completion(updated.InstanceId, cancellationToken);
        if (comp.AllCompleted && inst.Status != InstanceStatus.Completed)
        {
            inst.Status = InstanceStatus.Completed;
            inst.UpdatedAt = now;
            await _store.UpdateInstance(inst, cancellationToken);
        }
        return new CompleteResult { Task = updated, Completion = comp };
    }

    private async Task CanAct(WorkflowTask task, string actor, CancellationToken cancellationToken)
    {
        switch (task.AssigneeKind)
        {
            case AssigneeKind.User:
                if (task.AssigneeId != actor)
                    throw EngineException.Forbidden();
                break;
            case AssigneeKind.Group:
                if (!await _dir.IsMember(actor, task.AssigneeId, cancellationToken))
                    throw EngineException.Forbidden();
                break;
            default:
                throw EngineException.Forbidden();
        }
    }

    private async Task CanComplete(WorkflowTask task, string actor, CancellationToken cancellationToken)
    {
        await CanAct(task, actor, cancellationToken);
        if (task.AssigneeKind == AssigneeKind.Group)
        {
            if (task.Status != TaskStatus.Claimed || task.ClaimedBy != actor)
                throw EngineException.NotClaimed();
        }
        if (task.Status == TaskStatus.Claimed && task.ClaimedBy != actor)
            throw EngineException.Forbidden();
    }
}
